File: Liftr/LevelProgress.swift
import Foundation

func levelProgressRatio(
    totalXp: Int64,
    currentLevelThresholdXp: Int64,
    nextLevelThresholdXp: Int64
) -> Double {
    let span = nextLevelThresholdXp - currentLevelThresholdXp
    guard span > 0 else { return 0 }
    let earned = totalXp - currentLevelThresholdXp
    return min(1.0, max(0.0, Double(earned) / Double(span)))
}

File: android/app/src/main/java/com/lilru/liftr/domain/LevelProgress.kt
package com.lilru.liftr.domain

import kotlin.math.max
import kotlin.math.min

fun levelProgressRatio(
    totalXp: Long,
    currentFloor: Long,
    nextCap: Long
): Double {
    val span = nextCap - currentFloor
    if (span <= 0) return 0.0
    val earned = totalXp - currentFloor
    return min(1.0, max(0.0, earned.toDouble() / span.toDouble()))
}
